src/libraries/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Windows/CertificatePal.PrivateKey.cs

```diff
@@ -28,10 +28,12 @@ public bool HasPrivateKey
             return GetPrivateKey<RSA>(
                 delegate (CspParameters csp)
                 {
+                    Debug.Assert(OperatingSystem.IsWindows());
```

Comment:
Can we eliminate all of these asserts if we paint the whole class?  Largely I'm looking for the pattern that we want to do with new code, and all these asserts feel like they'd get old, fast.

Reply:
Byt the design MSBuild emitting `SupportedOSPlatform("windows")` attribute for the builds targeting windows, which is working well for normal projects and the analyzer wouldn't warn, not sure if/how that is/could applicable for the runtime projects. But it would be great if we could not warn for windows only builds in runtime

Reply:
Why is it needed? Could we make Pal.Windows to be included in Windows builds only?

Reply:
> Could we make Pal.Windows to be included in Windows builds only?

It is...
https://github.com/dotnet/runtime/blob/3be4a5719a3a78c2a1e60c928c10aa48f2da69fd/src/libraries/System.Security.Cryptography.X509Certificates/src/System.Security.Cryptography.X509Certificates.csproj#L231-L234

The only place I notice that we apply the OS-specificness to an assembly is https://github.com/dotnet/runtime/blob/196fd1804dcc36fc6a398bcc318e07a171a0fc9c/eng/versioning.targets#L25-L30, and the only assemblies that do that are always Windows-only.  Seems like we have a gap for our "Windows-specific implementation assembly" assemblies.

Reply:
From what i am seeing the windows targets of runtime projects are not the same as the standard one `net5.0-windows` (has suffix `_NT` like `<TargetFrameworks>$(NetCoreAppCurrent)-Windows_NT;</TargetFrameworks>`). That means MSBuild is not emitting the desired attribute. But i am not sure even fixing the suffix would fix the issue, do not know well how the runtime build works cc @safern @ViktorHofer

Reply:
> Could we make Pal.Windows to be included in Windows builds only?

Even it is included in Windows build it would  still warn

Reply:
Repro'ing in another project, using a tfm of `net5.0-windows` yields the expected behavior of not producing warnings for calling Windows-only methods.

> That means MSBuild is not emitting the desired attribute

@buyaa-n I forget which attribute it is that gets emitted when `net5.0-windows` is used, and where that is picked up. Can you remind me what property gets set by that and how the analyzer receives it and uses it?

Reply:
> Repro'ing in another project, using a tfm of net5.0-windows yields the expected behavior of not producing warnings for calling Windows-only methods.

Could you try with `net5.0-windows7.0`? as there is a [bug ](https://github.com/dotnet/sdk/issues/13494)that MSBuild expecting a version (which is fixed but might not deployed yet

>  I forget which attribute it is that gets emitted when net5.0-windows is used, and where that is picked up. Can you remind me what property gets set by that and how the analyzer receives it and uses it?

It should emit [SupportedOSPlatform("windows")] attribute in AssemblyInfo.cs file, which will have an assembly-level application of SupportedOSPlatformAttribute.

Reply:
just `net5.0-windows` without a version is working for me, i have installed 5.0 sdk RC2 a several days ago

Reply:
> The only place I notice that we apply the OS-specificness to an assembly is runtime/eng/versioning.targets
> 
>  , and the only assemblies that do that are always Windows-only. Seems like we have a gap for our "Windows-specific implementation assembly" assemblies.

Seems i can add there [SupportedOSPlatform("windows")] attribute for windows specific builds, thanks @bartonjs, it is solved at least libraries warnings, but not corelib. I am raising another PR for fix https://github.com/dotnet/runtime/pull/43097 and closing this draft